dids/did_dht.go

```diff
@@ -0,0 +1,322 @@
+	}
+
+	document := didRecord.DIDDocument(uri)
+	return ResolutionResultWithDocument(*document), nil
+}
+
+// parseDNSDID takes the bytes of the DNS representation of a DID and creates an internal representation
+// used to create a DID document
+func parseDNSDID(data []byte) (*dhtDIDRecord, error) {
+	var p dnsmessage.Parser
+	if _, err := p.Start(data); err != nil {
+		return nil, err
+	}
+
+	didRecord := dhtDIDRecord{
+		records: make(map[string]string),
+	}
+
+	// need to skip questions to move the index to the right place to read answers
+	if err := p.SkipAllQuestions(); err != nil {
+		return nil, err
+	}
+
+	for {
+		h, err := p.AnswerHeader()
+		if err == dnsmessage.ErrSectionDone {
+			break
+		}
+
+		if h.Type != dnsmessage.TypeTXT {
+			continue
+		}
+
+		value, err := p.TXTResource()
+		if err != nil {
+			// TODO check what kind of error and see if this should fail
+		}
+
+		name := h.Name.String()
+		fullTxtRecord := strings.Join(value.TXT, "")
+		if strings.HasPrefix(name, "_did") {
+			didRecord.rootRecord = fullTxtRecord
+			continue
+		}
+
+		if _, ok := didRecord.records[h.Name.String()]; !ok {
+
+		}
+		didRecord.records[h.Name.String()] = fullTxtRecord
+	}
+
+	return &didRecord, nil
+}
+
+// TODO on the diddhtrecord we should validate the minimum reqs for a valid did
+func parseVerificationRelationships(rootRecord string) map[string][]string {
+	rootRecordProps := parseTXTRecordData(rootRecord)
+
+	// reverse the map to get the relationships
+	relationshipMap := map[string][]string{}
+	for k, values := range rootRecordProps {
+		v := strings.Join(values, "")
+		rel, ok := relationshipMap[v]
+		if !ok {
+			rel = []string{}
+		}
+		rel = append(rel, k)
+		relationshipMap[v] = rel
+	}
+
+	return relationshipMap
+}
+
+func parseTXTRecordData(data string) map[string][]string {
+	result := map[string][]string{}
+	fields := strings.Split(data, ";")
+	for _, field := range fields {
+		kv := strings.Split(field, "=")
+		k, v := kv[0], strings.Split(kv[1], ",")
+		current, ok := result[k]
+		if ok {
+			v = append(current, v...)
+		}
+		result[k] = v
+	}
+
+	return result
+}
+
+func Create(did *Document) {
+
+}
+
+// UnmarshalVerificationMethod unpacks the TXT DNS resource encoded verification method
+func UnmarshalVerificationMethod(data string) (*VerificationMethod, error) {
+	propertyMap := parseTXTRecordData(data)
+
+	vm := &VerificationMethod{}
+	var key string
+	var algorithmID string
+	for property, v := range propertyMap {
+		switch property {
+		// According to https://did-dht.com/#verification-methods, this should not be a list
+		case "id":
+			vm.ID = strings.Join(v, "")
+		case "t": // Index of the key type https://did-dht.com/registry/index.html#key-type-index
+			algorithmID, _ = keyTypes[strings.Join(v, "")]
+		case "k": // unpadded base64URL representation of the public key
+			key = strings.Join(v, "")
+		case "c": // the controller is optional
+			vm.Controller = strings.Join(v, "")
+		}
```

Comment:
Should there be a `default:` case here that returns an error in case of an unsupported propertY?

Reply:
I'd go for skipping it and maybe logging something. if something new gets added to the spec, I don't know if this should completely fail. @mistermoe thoughts?